Assets/Scripts/Communication/CommunicationManager.cs

```diff
@@ -0,0 +1,392 @@
+						}
+						Debug.Log("Exception getting response: " + e.ToString());
+						throw e;
+					}
+					
+					if (resp == null){
+						break;
+					}
+
+					if (resp.ContainsKey("Request_ID") && (string) resp["Request_ID"] == requestID){
+						response = resp;
+					}
+					else{
+						lock(asyncMessagesLock){
+							Debug.Log("Found an async message");
+							LogDictionary(resp);
+							asyncMessagesQueue.Enqueue(resp);
+						}
+					}
+				}
+				
+				if (response != null){
+					//Add the response to the common dictionary
+					lock(responseDictLock){
+						responseDict[requestID] = response;
+					}
+				}
+				else if (request == null){
+					//No Request or response found, just continue looping
+				}
+				else{
+					//Unable to process request, re-add it to process it later
+					lock(requestQueueLock){
+						requestQueue.Enqueue(request);
+					}
+				}
+			}
+		} catch (Exception e){
+			Debug.Log("Manager Thread crashed: " + e.ToString());
+		}
+
+		
+	}
+
+	public static void OnDisable()
+	{
+		// If the thread is still running, we should shut it down,
+		if (_threadRunning)
+		{
+			// This forces the while loop of the thread to exit
+			_threadRunning = false;
+
+			// This waits until the thread exits
+			_manager.Join();
+		}
+
+		Disconnect();
+	}
+
+	/********************************************
+	 * Public Interface functions
+	 ********************************************/
+	public static string Request(Dictionary<string, object> data)
+	{
+		//Communication Manager Thread crashed, restart it
+		if (!_threadRunning){
+			CommunicationManager.Start();
+		}
+
+		//Generate a request id
+		Guid g = Guid.NewGuid();
+		data["Request_ID"] = g.ToString();
+
+		//Put the request id and the request data into the request Queue
+		lock (requestQueueLock)
+		{
+			requestQueue.Enqueue(data);
+		}
+
+		//Return the request id
+		return g.ToString();
+	}
+
+	public static Dictionary<string, object> RequestAndGetResponse(Dictionary<string, object> data){
+		string requestID = Request(data);
+
+		return GetResponse(requestID);
+	}
+
+	public static Dictionary<string, object> GetResponse(string request_id, bool blocking = true, int timeout = 2, int sleep_time = 100)
+	{
+		int elapsed_time = 0;
+		Dictionary<string, object> response = null;
+
+		if (!blocking)
+		{
+			lock (responseDictLock)
+			{
+				if (responseDict.ContainsKey(request_id))
+				{
+					response = responseDict[request_id];
+					responseDict.Remove(request_id);
+				}
+			}
+
+			return response;
+		}
+
+		while (elapsed_time < timeout)
+		{
+			lock (responseDictLock)
+			{
+				//Debug.Log("Checking if a response exists for request: " + request_id.ToString());
+				if (responseDict.ContainsKey(request_id))
+				{
+					response = responseDict[request_id];
+					responseDict.Remove(request_id);
+				}
+			}
+
+			if (response != null)
+			{
+				return response;
+			}
+			else
+			{
+				Thread.Sleep(sleep_time);
+				elapsed_time += sleep_time / 1000;
```

Comment:
I mean, want me to raise an Exception?  Or return null?  I would hope who ever wants to use this function for legitimate purposes wouldn't pass a negative number.  In testing, they should see it throw an exception if a negative sleep value is specified.